File: dotnet/Lexy.Poc/Lexy.Poc.Core/Parser/Tokens/Token.cs
using System;

namespace Lexy.Poc.Core.Parser.Tokens
{
    public abstract class Token : IToken
    {
        public TokenCharacter FirstCharacter { get; }

        public abstract string Value { get; }

        protected Token(TokenCharacter firstCharacter)
        {
            FirstCharacter = firstCharacter ?? throw new ArgumentNullException(nameof(firstCharacter));
        }
    }

    public interface IToken
    {
        TokenCharacter FirstCharacter { get; }
    }

    public class TokenCharacter
    {
        public int Position { get; }
        public char Value { get; }

        public TokenCharacter(char value, int position)
        {
            Value = value;
            Position = position;
        }
    }
}

File: typescipt/src/parser/tokens/token.ts
import {IToken} from "./IToken";
import {TokenCharacter} from "./tokenCharacter";

export abstract class Token implements IToken {

  abstract tokenIsLiteral: boolean;
  abstract tokenType: string;
  abstract value: string;

  public firstCharacter: TokenCharacter;

  protected constructor(firstCharacter: TokenCharacter) {
    this.firstCharacter = firstCharacter;
  }
}